[Data] Turbopuffer Datasink

nehiljain opened the pull request:
## Description

This PR adds a `TurbopufferDatasink` for Ray Data, enabling Ray datasets to be written directly into the Turbopuffer vector database. The datasink supports both single-namespace writes and multi-namespace writes, where rows are grouped by a `namespace_column` and written into separate Turbopuffer namespaces derived from that column.

The implementation includes:
- A `TurbopufferDatasink(Datasink)` that:
  - Validates configuration (mutually exclusive `namespace` vs `namespace_column`, required API key, namespace format).
  - Aggregates Ray blocks into a single PyArrow table, prepares it by renaming `id_column` → `"id"` and `vector_column` → `"vector"`, and filters out rows with null IDs.
  - Handles multi-namespace mode by grouping on `namespace_column`, formatting namespace names via `namespace_format` (e.g., `block_spans__{namespace}`), and writing each group to its own Turbopuffer namespace.
  - Converts bytes fields to stable, JSON-serializable representations:
    - 16-byte UUIDs → string UUIDs.
    - All other `bytes` (including inside lists) → hex strings.
- Safer schema handling in `_prepare_arrow_table`:
  - Raises a `ValueError` if renaming a custom `id_column` to `"id"` would conflict with an existing `"id"` column.
  - Raises a `ValueError` if renaming a custom `vector_column` to `"vector"` would conflict with an existing `"vector"` column.
  - This avoids ambiguous duplicate column names when later accessing `table.column("id")` / `"vector"`.

A comprehensive test suite is added/updated in `python/ray/data/tests/test_turbopuffer_datasink.py` to cover:
- Constructor validation (namespace config, namespace format placeholder, API key from param or env).
- Client initialization and region defaulting.
- Arrow table preparation:
  - Renaming and null-ID filtering.
  - Error on missing custom ID column.
  - Errors when renaming would produce duplicate `"id"` or `"vector"` columns.
- Multi-namespace grouping behavior (including UUID-typed namespace keys) and error on missing `namespace_column`.
- Batching behavior in single-namespace mode.
- Transform-to-Turbopuffer-row behavior:
  - Requires an `"id"` column.
  - Correctly converts UUID bytes and non-UUID bytes (including inside lists) to consistent string/hex representations.
- Retry behavior:
  - Success on first attempt.
  - Transient failures that eventually succeed, verifying backoff is invoked.
  - Persistent failures that exhaust retries and raise.
- End-to-end write orchestration in both single-namespace and multi-namespace modes.

This PR is aligned with the design and performance considerations described in `turbopuffer_datasink_spec.md`, including support for multi-tenant (multi-namespace) ingestion patterns and Turbopuffer’s performance guidance (schema types, batch sizing, concurrency).


Test Script:
```
"""Larger-scale test for write_turbopuffer."""

import time
import numpy as np
import ray

# Configuration
API_KEY = "..."
NUM_ROWS = 100_000
VECTOR_DIM = 384  # Common embedding dimension (e.g., sentence-transformers)
NUM_PARTITIONS = 10

# Initialize Ray
ray.init()

print(f"Generating {NUM_ROWS:,} rows with {VECTOR_DIM}-dim vectors...")


def generate_batch(batch):
    """Generate random vectors and metadata for each row."""
    n = len(batch["id"])

    # Vectors must be list-of-lists for Turbopuffer JSON serialization
    vectors = np.random.randn(n, VECTOR_DIM).astype(np.float32)

    return {
        "id": batch["id"],
        "vector": [v.tolist() for v in vectors],
        "category": np.char.add("cat-", (batch["id"] % 100).astype(str)),
        "score": np.random.rand(n).astype(np.float32),
    }


# Create dataset: range -> map to add vectors and attributes
ds = (
    ray.data.range(NUM_ROWS)
    .repartition(NUM_PARTITIONS)
    .map_batches(generate_batch, batch_size=1000)
)

# Write to Turbopuffer
print("Writing to Turbopuffer...")
start = time.time()

ds.write_turbopuffer(
    namespace="ray-scale-test",
    api_key=API_KEY,
    region="...",
    schema={
        "score": {"type": "float"},
        "category": {"type": "string"},
    },
)

elapsed = time.time() - start
print(
    f"Done! Wrote {NUM_ROWS:,} vectors in {elapsed:.1f}s "
    f"({NUM_ROWS / elapsed:.0f} rows/sec)"
)
```

## Reference docs

- Turbopuffer docs: https://turbopuffer.com/docs
- Turbopuffer performance guide: https://turbopuffer.com/docs/performance

nehiljain committed:
turbopuffer datasink with unit testing

nehiljain committed:
Fix Turbopuffer datasink bytes handling and column rename collisions

nehiljain committed:
Fix Turbopuffer multi-namespace null namespace bug

nehiljain committed:
fixed and comment resolutions

nehiljain committed:
data: test Turbopuffer namespace_column rename

nehiljain committed:
Make Turbopuffer distance metric configurable

nehiljain committed:
Optimize Turbopuffer multi-namespace grouping

nehiljain committed:
Refactor Turbopuffer datasink improvements

- Use Ray's standard _check_import utility for consistent error messages
- Add _rename_single_column helper to reduce duplication
- Check table schema before to_pylist conversion for efficiency
- Fix retry jitter calculation to use full jitter (0-100% range)

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,489 @@
+        3. Filter out rows with null IDs
+        """
+        # Rename ID column if needed
+        if self.id_column != "id":
```
Can you please add this in the comments?

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,496 @@
+
+        return table
+
+    def _rename_single_column(
```
This method can be removed.

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,496 @@
+        num_rows = len(table)
+        for batch_start in range(0, num_rows, self.batch_size):
+            batch_end = min(batch_start + self.batch_size, num_rows)
+            batch_table = table.slice(batch_start, batch_end - batch_start)
```
```
for batch in table.to_batches(max_chunksize=self.batch_size):
    batch_table = pa.Table.from_batches([batch])
```
is a bit more readable

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,496 @@
+
+        return rows
+
+    def _write_batch_with_retry(
```
```
from ray._common.retry import call_with_retry

def _write_batch_with_retry(
    self,
    namespace,
    batch_data: list,
    namespace_name: str,
):
    """
    Write a single batch with exponential backoff retry using Ray's common utility.
    """
    def _write_fn():
        namespace.write(
            upsert_rows=batch_data,
            schema=self.schema,
            distance_metric=self.distance_metric,
        )
        return  # Success

    try:
        call_with_retry(
            _write_fn,
            description=f"write batch to namespace '{namespace_name}'",
            max_attempts=5,
            max_backoff_s=32.0,
        )
        logger.debug(
            f"Successfully wrote {len(batch_data)} rows to namespace '{namespace_name}'"
        )
    except Exception as e:
        logger.error(
            f"Write failed for namespace '{namespace_name}' after retries: {e}"
        )
        raise
```

Let's also add the appropriate type annotation for namespace

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,448 @@
+            namespace_name = self.namespace_format.format(namespace=namespace_str)
+
+            # Write this group
+            self._write_single_namespace(client, group_table, namespace_name)
```
+1

nehiljain committed:
Add type annotation for namespace parameter in TurbopufferDatasink

Added turbopuffer.Namespace type annotation to improve IDE support and type checking for the namespace parameter in _write_batch_with_retry method.

nehiljain committed:
Refactor batching logic in TurbopufferDatasink to use PyArrow's to_batches

Replace manual index-based slicing with table.to_batches(max_chunksize) for improved readability and idiomatic PyArrow usage.

nehiljain committed:
Use defaultdict for namespace grouping in TurbopufferDatasink

Replace dict with setdefault() with defaultdict(list) for cleaner, more idiomatic code when building namespace index groups.

nehiljain committed:
Simplify null ID filtering in TurbopufferDatasink with pc.is_valid

Replace verbose pc.invert(pc.is_null()) pattern with idiomatic PyArrow pc.is_valid() function for cleaner null filtering.

nehiljain committed:
Remove custom _rename_single_column helper in TurbopufferDatasink

Replace with PyArrow's built-in rename_columns() method which accepts a dict mapping old to new column names. Removes 15 lines of unnecessary helper code.

nehiljain committed:
Optimize multi-namespace grouping with PyArrow group_by and sorting

Replace manual index grouping with PyArrow's native sort + group_by().aggregate() + slice approach. Leverages vectorized C++ operations and efficient slicing for better performance with many namespaces.

nehiljain committed:
Fix TurbopufferDatasink serialization with __getstate__/__setstate__

Exclude both _client and _turbopuffer from pickling to avoid
serialization issues when distributing to worker nodes. Re-import
turbopuffer module in __setstate__.

nehiljain committed:
Add serialization test for TurbopufferDatasink

Verifies that __getstate__ and __setstate__ properly handle
non-serializable attributes (_turbopuffer module and _client object).
This ensures the datasink can be pickled/unpickled correctly when
Ray sends it to worker nodes.

nehiljain commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,504 @@
+                    f"'{self.id_column}' to 'id'. Please disambiguate your schema."
+                )
+
+            table = table.rename_columns({self.id_column: "id"})
```
This is not true PyArrow 19.0.1 does support dict.

nehiljain committed:
Use Ray's common retry utility in TurbopufferDatasink

Replace custom retry logic with call_with_retry for consistency with Ray patterns.
Simplifies code by using binary exponential backoff with jitter from ray._common.retry.

nehiljain committed:
refactor(turbopuffer): DRY column renaming with helper method and constants

- Add _ID_COLUMN and _VECTOR_COLUMN constants for reserved column names
- Extract _rename_column_if_needed() helper to eliminate duplicate logic
- Update _prepare_arrow_table() and _write_multi_namespace() to use constants
- Update test regex patterns for more flexible error matching

nehiljain committed:
fix(turbopuffer): sort aggregate result to prevent namespace mismatch

PyArrow's group_by().aggregate() with use_threads=True (default) does NOT
guarantee output order matches input order. The previous code assumed the
aggregate result order matched the sorted table order, which could cause
rows to be assigned to wrong namespaces (silent data corruption).

Fix: explicitly sort the aggregate result by the group column to ensure
it matches the sorted table order before slicing.

nehiljain commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,496 @@
+        new_names[idx] = dst_name
+        return table.rename_columns(new_names)
+
+    def _write_multi_namespace(self, client, table: pa.Table):
```
Turbopuffer requires each write call to target a single namespace but your dataset contains records destined for many namespaces.
Notion's Use Case: Their embedding pipeline processes data from hundreds of workspaces in a single batch.

alexeykudinkin replied:
Yeah, but they should do repartitioning by namespace before writing, rather than doing N calls from every task

nehiljain commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,496 @@
+                        else:
+                            # Keep non-bytes items as-is
+                            converted_list.append(item)
+                    row[key] = converted_list
```
Based on current schema we are looking at its flat so I am assuming all attribute columns become top-level key-value attributes. Do you think we should make it more general? I feel its complexity for no reason. Keeping it flat forces the client to handle that complexity

goutamvenkat-anyscale replied:
The current version is rather hard to parse imo. 

Actually can you explain why bytes needs to be converted to a 16 byte UUID?

nehiljain committed:
Fix inconsistent non-UUID bytes handling in namespace column

Non-UUID bytes in namespace column values were converted using str()
which produces Python repr like "b'\x01\x02\x03'", while the same
bytes in _transform_to_turbopuffer_format used .hex() producing clean
hex strings like "010203". This inconsistency meant the same byte
sequence would be represented differently depending on context.

Fixed by using .hex() for non-UUID bytes in namespace column handling,
matching the behavior in _transform_to_turbopuffer_format.

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,512 @@
+        self.namespace = namespace
+        self.namespace_column = namespace_column
+        self.namespace_format = namespace_format
+        self.api_key = api_key or os.getenv("TURBOPUFFER_API_KEY")
```
Curious why we allow both options?

nehiljain replied:
No specific reason, just flexibility for the client to use it however they want it to. What do you think is the best way? If we had to choose one may be we should let them set it like `self.api_key = api_key` this.

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,512 @@
+                f"'{source_column}' to '{target_column}'. Please disambiguate your schema."
+            )
+
+        return table.rename_columns({source_column: target_column})
```
Can we use `arrow_block.py`'s abstractions for renaming columns?

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,512 @@
+        start_idx = 0
+        for i in range(num_groups):
+            namespace_value = group_keys[i].as_py()
+            count = group_counts[i].as_py()
```
```
from ray.data._internal.arrow_block import ArrowBlockAccessor, ArrowBlockColumnAccessor

block_accessor = ArrowBlockAccessor(sorted_table)
group_col = sorted_table.column(group_col_name)
col_accessor = ArrowBlockColumnAccessor(group_col)

counts_dict = col_accessor.value_counts()

if counts_dict:
    groups = sorted(
        zip(counts_dict["values"], counts_dict["counts"]),
        key=lambda x: x[0]
    )

    num_groups = len(groups)

    start_idx = 0
    for namespace_value, count in groups:
        group_table = block_accessor.slice(start_idx, start_idx + count)
        start_idx += count
```

goutamvenkat-anyscale reviewed (changes requested):
Some more comments. Thanks

nehiljain committed:
Use arrow_block.py abstractions for sort and rename in TurbopufferDatasink

nehiljain committed:
fix(turbopuffer): stream blocks individually to avoid heap blowup

nehiljain committed:
fix(data): simplify TurbopufferDatasink UUID handling for better performance

nehiljain committed:
refactor(data): Use Ray utilities for group-by-slice in TurbopufferDatasink

Replace manual group_by().aggregate() pattern with Ray Data utilities:
- ArrowBlockColumnAccessor.value_counts() for group counting
- ArrowBlockAccessor.slice() for slicing with (start, end) semantics

This reduces raw PyArrow API calls and follows patterns used elsewhere
in the Ray Data codebase.

goutamvenkat-anyscale commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,487 @@
+        """Lazy initialize Turbopuffer client"""
+        if self._client is None:
+            # Initialize with api_key and region (default to "gcp-us-central1" if not specified)
+            region = self.region or "gcp-us-central1"
```
Let's make this a const

goutamvenkat-anyscale reviewed (changes requested):
Just one coment

alexeykudinkin commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,487 @@
+
+        import turbopuffer
+
+        self._turbopuffer = turbopuffer
```
What's this for?

alexeykudinkin commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,487 @@
+        )
+        table = self._rename_column_if_needed(
+            table, self.vector_column, _VECTOR_COLUMN, "Vector"
+        )
```
- Project table to only include these 2 columns first
 - Rename them to the expected names
 - Then you can remove this util

goutamvenkat-anyscale replied:
Yes nehil is correct, we can't project first then rename. That would drop other columns.

alexeykudinkin commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,487 @@
+                    # with _transform_to_turbopuffer_format
+                    namespace_str = namespace_value.hex()
+            else:
+                namespace_str = str(namespace_value)
```
Use `BlockAccessor._iter_groups_sorted` doing that for you

nehiljain replied:
I couldnt find this func `_iter_groups_sorted`? Do we want to implement a new utility in BlackAccessor for this?

goutamvenkat-anyscale replied:
@alexeykudinkin it's not implemented in OSS

alexeykudinkin replied:
Then we'd just implement it:

```
    def _iter_groups_sorted(
        self, sort_key: "SortKey"
    ) -> Iterator[Tuple[Sequence[KeyType], Block]]:
        """
        NOTE: THIS METHOD ASSUMES THE BLOCK BEING SORTED

        Creates an iterator over (zero-copy) blocks of rows grouped by
        provided key(s).
        """

        key_col_names: List[str] = sort_key.get_columns()

        if not key_col_names:
            # Global aggregation consists of a single "group", so we short-circuit.
            yield tuple(), self.to_block()
            return

        boundaries = self._get_group_boundaries_sorted(key_col_names)

        for start, end in zip(boundaries[:-1], boundaries[1:]):
            # Fetch tuple of key values from the first row
            row = self._get_row(start)

            yield row[key_col_names], self.slice(start, end, copy=False)

```

nehiljain commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,487 @@
+            and self.namespace_column == self.id_column
+            and _ID_COLUMN in table.column_names
+        ):
+            group_col_name = _ID_COLUMN
```
> Why is namespace column equal to id column?

I think you are right, this makes no sense. We wouldnt want namespace be 1:1 on vectors/docs in it. If that is not true we can remove that piece and just tell the user that we don't support the same col to be used by both

nehiljain committed:
Resolved various comments to make the code simpler and performant.
  - Make region parameter required (was optional with default)
  - Add validation: namespace_column cannot equal id_column or vector_column
  - Switch from row-oriented (upsert_rows) to column-oriented (upsert_columns) format
  - Remove intermediate pa.Table allocation in write loop
  - Simplify _write_multi_namespace by removing column name resolution logic
  - Update docstring to document namespace_column constraints

goutamvenkat-anyscale commented on python/ray/data/dataset.py:
```diff
@@ -5256,6 +5257,131 @@ def write_clickhouse(
+            namespace_column=namespace_column,
+            namespace_format=namespace_format,
+            api_key=api_key,
+            region=region,
```
+1

goutamvenkat-anyscale committed:
Merge from master

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Made some fixes

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Region is needed

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Linter

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
only run on pyarrow >= 19.0

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
remove file

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Fix in rename cols

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Clean up

Signed-off-by: Goutam <goutam@anyscale.com>

alexeykudinkin commented on python/ray/data/_internal/datasource/turbopuffer_datasink.py:
```diff
@@ -0,0 +1,460 @@
+        """
+        Write table to a single namespace with batching.
+        """
+        logger.debug(f"Writing {len(table)} rows to namespace '{namespace_name}'")
```
Let's nuke

goutamvenkat-anyscale committed:
Remove multi namespace writes + address comments

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
clean up

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Remove dead code

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
add multinamespace writes back... no custom formatting

Signed-off-by: Goutam <goutam@anyscale.com>

goutamvenkat-anyscale committed:
Merge branch 'master' into feat-turbopuffer-datasink
